### PR TITLE
Adding a check for missing ticks

### LINKED ISSUE #84: Check for missing ticks / empty tick data
We want a hard check on possible missing ticks in downloaded data. This means that we have to check if all downloaded pairs have data for every needed tick. It could be that f.e. BTC/USDT has 1 missing tick on 13.00, while the other downloaded pairs do actually have this tick. In that case, we want to be notified using console error output. 

If you have any questions, hit us up on Discord.

### PR DESCRIPTION
**Pull request by pvdklei**
# Fixes

Fixes #84

# What has been changed?

In `Datamodule` I added a method that checks for missing ticks. If more than zero prints a warning. I did not find missing data, either by looking at the data or by this method, so it does not show anything right now. Warning what exact dates/ticks are missing seems too much, maybe when after adding a `--verbose` option in the cli. 

# Examples

`[WARNING] Pair 'USDT/BITC' contains 4 missing ticks`

# Actions to be performed before this can be merged

- [ ] PR depends on #
- [ ] Script to be run

# Security Measures

# Potential Issues/Future considerations

# Additional Info

### CODE REVIEW

**Comment by keebrev** [comment]
@pvdklei did you encounter missing ticks while testing this for this issue?

**Comment by pvdklei** [comment]
@keebrev like I said, no. At least no null/NaN values. It could be that the exchange library you use already filters those out, and that indices (dates) are missing though.

**Comment by dema-trading-ai** [comment]
> @keebrev like I said, no. At least no null/NaN values. It could be that the exchange library you use already filters those out, and that indices (dates) are missing though.

 In case we configured a 5 minute timeframe, ALL ticks should be 5m apart from eachother. Could you add another check to validate if this is the case for every candle?

**Comment by noudcorten** [comment]
@dema-trading-ai @pvdklei I think a simple check is to create a list containing all the ticks that the dataframe should have (look at datamodule.py -> download_data_for_pair). Then compare this list with the index of the dataframe to see if there are some ticks missing. Simply said do something like:
```python
required_ticks = np.arange(start_date, data_to, self.timeframe_calc)
missing_ticks = set(required_ticks) - set(df.index)
```
Then, if len(missing_ticks) > 0 you know there is a tick missing (and which one)

**Comment by pvdklei** [comment]
@keebrev I added something like @noudcorten suggested, should be oke now!

**Comment by noudcorten** [comment]
@pvdklei Looks nice! But maybe it's nicer to first find out which method binance uses to fill up it's empty candles and then build a check around that. It could be on of these scenario's:
A. Empty candles are returned as timestamps with NaN as Open, High, Low, Close and Volume.
B. Empty candles are not returned and are missing in the data
C. There are no empty candles

So find which is scenario is true and then build the check around that, instead of covering it all.

**Comment by keebrev** [comment]
@noudcorten I agree with you, although, since we do not know IF there is missing data and if there is, HOW we can find it, I think a cover all is doing the job for now. As soon as we know how missing data could be traced, we can update this code.

**Comment by noudcorten** [comment]
@pvdklei @dema-trading-ai This PR is outdated. Index of the dataframe is now just the timestamp in miliseconds as an integer. So we're not working with datetime as index anymore.

**Comment by pvdklei** [comment]
@noudcorten ccxt is unclear about what they do with missing values, so best we can do is guess that they add null values or delete rows. If they replaced missing values with something else than 'not a number', for instance with the mean, there is no way to find out which those were. 

Btw I changed it so it fit the new timestamp system

**Comment by noudcorten** [comment]
@pvdklei Yes I agree with you! Then it's indeed better to keep it this way.

File: data/datamodule.py [hunk 1]
```diff
@@ -339,3 +342,30 @@ def remove_backtesting_file(self, pair: str) -> None:
...
+
+            # any NaN?
+            n_nan = data.isnull().any(axis="columns").sum()
+            
+            # missing dates?
+            index = data.index.to_numpy().astype(int)
+            diff = np.setdiff1d(daterange, index)
+            n_missing = len(diff)
+
+            if n_nan > 0 or n_missing > 0:
```
Comment (noudcorten) [inline data/datamodule.py:364]: Maybe change this to two if-statements, so that when one of the two triggers the print statement doesn't say '0 ...'.

Comment (pvdklei) [reply data/datamodule.py:364]: Very true

File: data/datamodule.py [hunk 2]
```diff
@@ -339,3 +342,30 @@ def remove_backtesting_file(self, pair: str) -> None:
...
+                              self.backtesting_to,
+                              self.timeframe_calc)
+
+        for pair, data in self.history_data.items():
+
+            # any NaN?
+            n_nan = data.isnull().any(axis="columns").sum()
+            
+            # missing dates?
+            index = data.index.to_numpy().astype(int)
```
Comment (noudcorten) [inline data/datamodule.py:360]: .astype(int) can be removed here

Comment (pvdklei) [reply data/datamodule.py:360]: Thought so as well, but results in weird warning and longer running time (for some reason)

Comment (noudcorten) [reply data/datamodule.py:360]: Aah okay. Keep it then :)

**Review by noudcorten — APPROVED** [review]
Looks great! Solve codacy problem, then it's ok!

**Comment by pvdklei** [comment]
@noudcorten done

**Review by keebrev — APPROVED** [review]

**Pull request merged**